src/lib/operators/join_hash.cpp

```diff
@@ -275,35 +297,77 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
 
     CurrentScheduler::wait_for_tasks(jobs);
 
+    // (Hacky) short cut for AntiNullAsTrue
+    //          If there is any NULL value on the left side, do not bother probe as no tuples can be emitted
+    //          anyway. Doing this early out here is hacky, but during probing we assume NULL values on the left
+    //          side do not matter, so we'd have no chance detecting a NULL value on the left side there.
```

Comment:
I don't get this. Could you give an example?

Reply:
Okay, this is the hard question. Let me answer here first and once we find something that explains it properly I'll update the comment.

`T1 AntiNullAsTrue JOIN T2 ON T1.a = T2.a` is equivalent to `T1.a NOT IN (SELECT T2.a FROM T2)`.  [The Internet is full with explanations on why NOT IN never returns any tuples as soon as its subquery contains just a single NULL](https://www.tech-recipes.com/rx/57894/sql-server-three-valued-logic-not-clause-null-values/). It's a bit tricky to understand, but consistent across data bases. 

Since the JoinHash swaps input sides for Semi/Anti joins, the `T1` in the example is the "right" table from the JoinHash's perspective and the `T2` in the example is the "left" table. I.e., `T1` is the probe table and `T2` is the build table. The way JoinHash is build it simply discards null values on the build side because all other join modes do not care about them. AntiNullAsTrue is the only one that needs them - because for AntiNullAsTrue `T1.a = T2.b` for `T2.b == NULL` is interpreted as `TRUE` which causes AntiNullAsTrue to drop each individual tuple in `T2`. 

Was that useful?

Reply:
Got it. I think I had some issues with the "AntiNullAsTrue", where the "true" is negated by the Anti. Maybe "AntiNullsIgnored" and "AntiNullsReject" (or similar)?

Do we have a comment on their semantics with examples at the place where this is defined?

Also, I find "left and right side" confusing as these might be swapped. Furthermore, I would traditionally think of an example such as `1 IN (...)`, where "left side" now means my list. After the swap, wouldn't it be better to talk about "hashed and probed sides"? @Bouncner?

Reply:
> Maybe "AntiNullsIgnored" and "AntiNullsReject" (or similar)?

The naming is hard here, and I am not 100% happy with AntiNullAsTrue/False either. But at least, imo, those are precise: "Imagine an anti join as you know it with distinction on how it interprets NULL-bool values". 
"AntiNullsIgnored" and "AntiNullsReject" could mean anything, depending on how you read it. Will nulls on the inner side be "ignored/rejected"? Or does  "ignored/rejected" refer to the result of the join predicate? If so, how do "reject" a bool? 

> Do we have a comment on their semantics with examples at the place where this is defined?

This is the doc of the JoinMode. I had hoped this explains `AntiNullAsTrue` precisely.

```c++
// Let R and S be two tables and we want to perform `R <JoinMode> S ON <condition>`
// AntiNullAsTrue:    If for a tuple Ri in R, there is a tuple Sj in S so that <condition> is NULL or TRUE, Ri is
//                      dropped. This behavior mirrors NOT IN.
// AntiNullAsFalse:   If for a tuple Ri in R, there is a tuple Sj in S so that <condition> is TRUE, Ri is
//                      dropped. This behavior mirrors NOT EXISTS
```

> After the swap, wouldn't it be better to talk about "hashed and probed sides"? 

Find the left/right naming confusing as well, "build" and "probe" would probably be better.

Reply:
Ok. I was on mobile and didn't find the quoted comment. That should be sufficient for someone trying to work their way through this very special behavior.

I'll open an issue for the left/right thing.